subprojects/dependency-lock/src/main/java/org/gradle/internal/dependencylock/DefaultDependencyLockState.java

```diff
@@ -0,0 +1,72 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.dependencylock;
+
+import org.gradle.api.Action;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.ResolvableDependencies;
+import org.gradle.api.artifacts.component.ComponentSelector;
+import org.gradle.api.artifacts.component.ModuleComponentSelector;
+import org.gradle.api.artifacts.result.DependencyResult;
+import org.gradle.api.artifacts.result.ResolvedDependencyResult;
+import org.gradle.internal.dependencylock.model.DependencyLock;
+import org.gradle.internal.dependencylock.model.DependencyVersion;
+import org.gradle.internal.dependencylock.model.ModuleKey;
+
+public class DefaultDependencyLockState implements DependencyLockState {
+
+    private final DependencyLock dependencyLock = new DependencyLock();
+
+    @Override
+    public void resolveAndPersist(final Project project, final Configuration configuration) {
+        if (configuration.isCanBeResolved()) {
+            configuration.getIncoming().afterResolve(new Action<ResolvableDependencies>() {
+                @Override
+                public void execute(ResolvableDependencies resolvableDependencies) {
+                    resolvableDependencies.getResolutionResult().allDependencies(new Action<DependencyResult>() {
+                        @Override
+                        public void execute(DependencyResult dependencyResult) {
+                            if (dependencyResult instanceof ResolvedDependencyResult) {
+                                ResolvedDependencyResult resolvedDependencyResult = (ResolvedDependencyResult) dependencyResult;
+                                ComponentSelector requested = dependencyResult.getRequested();
+
+                                if (requested instanceof ModuleComponentSelector) {
+                                    ModuleComponentSelector requestedModule = (ModuleComponentSelector) requested;
+                                    addDependency(project.getPath(), configuration.getName(), requestedModule, resolvedDependencyResult, dependencyLock);
+                                }
+                            }
+                        }
+                    });
+                }
+            });
+        }
+    }
+
+    private void addDependency(String projectPath, String configurationName, ModuleComponentSelector requestedModule, ResolvedDependencyResult resolvedDependencyResult, DependencyLock dependencyLock) {
```

Comment:
Added. Maybe I misunderstood our discussions but I thought we landed at:

> Dependency metadata rules and dependency resolution rules can change over time. These rules take precedence over locked versions and are honored when they are changed by the user.

From what I understood we didn't want to lock resolve rules. The rules in the build script take precedence.

Reply:
How do you think this will work? Would we inject the locked version into a build _before_ running any rules? In that case, shouldn't we also be _reading_ the locked version before running any rules?

Currently, the version that will written to the lock file will be the one after all rules have been executed.